### ISSUE TITLE
semigroups 4.0.0 on cygwin

### ISSUE THREAD

**Issue by ChrisJefferson** [libsemigroups/libsemigroups#335]
semigroups 4.0.0 doesn't build on cygwin. This is because the 'erase' function on 'std::string' doesn't accept const_iterators (there is a boring story why this is true I won't write down unless someone cares. This is against the C++ standard).

The following patch fixes it (this is against the 4.0.0 release of semigroups, I wasn't positive which branch to apply this to)

```
diff -r orig/semigroups-4.0.0/libsemigroups/include/libsemigroups/kambites.hpp semigroups-4.0.0/libsemigroups/include/libsemigroups/kambites.hpp
66a67
>
413c414
<             y.erase(y.cbegin() + x.size(), y.cend());
---
>             y.erase(y.begin() + x.size(), y.end());
642c643
<           wp.erase(wp.cbegin(), wp.cbegin() + Z(s).size());
---
>           wp.erase(wp.begin(), wp.begin() + Z(s).size());
840c841
<       u.erase(u.cbegin(), u.cbegin() + Z(j).size());
---
>       u.erase(u.begin(), u.begin() + Z(j).size());
845c846
<       w.erase(w.cbegin() + i, w.cend());
---
>       w.erase(w.begin() + i, w.end());
946c947
<             u.erase(u.cbegin(), u.cbegin() + XYZ(i).size());
---
>             u.erase(u.begin(), u.begin() + XYZ(i).size());
950c951
<             v.erase(v.cbegin(), v.cbegin() + XYZ(j).size());
---
>             v.erase(v.begin(), v.begin() + XYZ(j).size());
960,961c961,962
<             u.erase(u.cbegin(), u.cbegin() + XY(i).size());
<             v.erase(v.cbegin(), v.cbegin() + XY(i).size());
---
>             u.erase(u.begin(), u.begin() + XY(i).size());
>             v.erase(v.begin(), v.begin() + XY(i).size());
963,964c964,965
<             u.erase(u.cbegin(), u.cbegin() + XY(i).size());
<             v.erase(v.cbegin(), v.cbegin() + XYZ(j).size());
---
>             u.erase(u.begin(), u.begin() + XY(i).size());
>             v.erase(v.begin(), v.begin() + XYZ(j).size());
968c969
<             u.erase(u.cbegin(), u.cbegin() + XYZ(i).size());
---
>             u.erase(u.begin(), u.begin() + XYZ(i).size());
970c971
<             v.erase(v.cbegin(), v.cbegin() + XY(j).size());
---
>             v.erase(v.begin(), v.begin() + XY(j).size());
984,985c985,986
<               u.erase(u.cbegin(), u.cbegin() + XYZ(i).size() - p.size());
<               v.erase(v.cbegin(), v.cbegin() + XYZ(j).size() - p.size());
---
>               u.erase(u.begin(), u.begin() + XYZ(i).size() - p.size());
>               v.erase(v.begin(), v.begin() + XYZ(j).size() - p.size());
```

**Comment by ChrisJefferson** [comment]
Actually, it should compile :)

The `erase` function was changed in C++11 to take `const_iterator` instead of `iterator`, and libsemigroups requires c++14.

This, and a number of other changes were made to std::string, which would make it impossible to link together programs that use the old (C++98) and new (C++11) std::string. Most linux distributions, at one of their major releases since 2011 "flipped the switch", and started using the new std::string implementation.

However, windows (and therefore cygwin) don't work like that -- people expect windows programs to work, and link together, forever. Therefore cygwin has turned on all new C++ features except those which would break linking to old programs (which is, in practice, some things in `std::string` and some obscure bits of exception handling you are unlikely to ever come across).

**Comment by james-d-mitchell** [comment]
Resolved in v2.1.4
